src/forms/create_ticket.ts

```diff
@@ -1,61 +1,116 @@
-                {
-                    name: 'priority',
-                    modal_label: 'Priority',
-                    type: AppFieldTypes.STATIC_SELECT,
-                    options: makeOptions(['urgent', 'high', 'normal', 'low']),
-                    is_required: false,
-                },
-                {
-                    name: 'additional_message',
-                    modal_label: 'Optional message',
-                    type: AppFieldTypes.TEXT,
-                    description: 'Add additional message to the Zendesk ticket',
-                    subtype: 'textarea',
-                    min_length: 2,
-                    max_length: formTextAreaMaxLength,
-                },
-                {
-                    name: 'post_message',
-                    modal_label: 'Mattermost message',
-                    type: AppFieldTypes.TEXT,
-                    value: postMessage,
-                    subtype: 'textarea',
-                    min_length: 2,
-                    max_length: formTextAreaMaxLength,
-                    is_required: true,
-                },
-            ],
+    const userID = call.context.acting_user_id || '';
+    const token = oauthStore.getToken(userID);
+    if (!token) {
+        throw new Error('Failed to get user access_token');
+    }
+    const zdClient = zd.newZDClient(token);
+    const formFields = await getFormFields(zdClient, call, postMessage);
 
+    const form: AppForm = {
+        title: 'Create Zendesk Ticket',
+        header: 'Create a Zendesk ticket from Mattermost by filling out and submitting this form. Additional text can be added in the `Optional Message` field.',
+        footer: 'Message modal form footer',
+        icon: zdIcon,
+        fields: formFields,
+        call: {
+            url: routes.app.BindingPathCreateForm + '?call_type=submit',
         },
     };
-    return callResponse;
+    return form;
+}
+
+// getFormFields returns a list of viewable app fields that have been mapped
+// from Zendesk form fields
+async function getFormFields(zdClient: Client, call, postMessage: string): Promise<AppField[]> {
+    const zdTicketForms = await tryCallWithMessage(zdClient.ticketforms.list(), 'Failed to fetch ticket forms');
+
+    let formID: number;
+    if (call.type === AppCallTypes.FORM && call.values.name === 'formName') {
+        formID = call.values.value.value;
+    }
+
+    const formField = mapFormsToSelectionField(zdTicketForms, formID);
+    const appFields = [formField];
+    if (!formID) {
+        return appFields;
+    }
+
+    const zdFormFieldIDs = getTicketFormFieldIDs(zdTicketForms, formID);
+    const zdTicketFields = await tryCallWithMessage(zdClient.ticketfields.list(), 'Failed to fetch ticket fields');
+    const zdViewableFields = getViewableFields(zdTicketFields, zdFormFieldIDs);
+    const formFields = mapFieldsToAppFields(zdViewableFields);
+
+    appFields.push(...formFields);
+    const postFields = getPostFormFields(postMessage);
```

Comment:
move these out of here? Need to test that values aren't lost when changing form types

Reply:
done.  Moved to field_mapper.

Implemented getSavedValue() to save the values when changing forms